File: app/src/main/java/com/ubudu/sample/model/Counter.java
package com.ubudu.sample.model;

import android.arch.lifecycle.MutableLiveData;

import java.util.Timer;
import java.util.TimerTask;

/**
 * Created by Aga on 02.02.2018.
 */

public class Counter {

    MutableLiveData <Integer> counter =  new MutableLiveData<>();

    public MutableLiveData<Integer> getCounter(){
        return counter;
    }

    public void startCounter(){

        //aga: zawsze z main thread'a
        counter.setValue(999);

        Timer start = new Timer();
        start.schedule(new TimerTask() {
            @Override
            public void run() {
                for(int i=0; i<2000; i++){
                    //aga: zawsze z worker thread'a
                    counter.postValue(i);
                    try {
                        Thread.sleep(1000);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
            }
        }, 2000);

    }


}

File: app/src/main/java/com/ubudu/sample/model/Counter.kt
package com.ubudu.sample.model

import java.util.Timer
import java.util.TimerTask

/**
 * Created by Aga on 02.02.2018.
 */

class Counter(counterMaxValue: Int) {

    var listener: CounterListener? = null

    private var counter = 0
    private var counterMaxValue = 0
    private var counterTimer = Timer()
    private var counterTimerTask: TimerTask? = null

    init {
        this.counterMaxValue = counterMaxValue
    }

    /**
     *
     */
    fun start() {
        // set initial value
        setValue(0)
        // new timer task
        counterTimerTask = object : TimerTask() {
            override fun run() {
                if (counter < counterMaxValue)
                    setValue(++counter)
                else
                    setValue(0)
            }
        }
        // new Timer
        counterTimer = Timer()
        // schedule timer task
        counterTimer.scheduleAtFixedRate(counterTimerTask, 1000, 1000)
    }

    /**
     *
     */
    fun stop() {
        counterTimerTask?.cancel()
        setValue(0)
    }

    /**
     *
     */
    private fun setValue(newValue: Int) {
        counter = newValue
        listener?.onValueChanged(counter)
    }

    interface CounterListener {
        fun onValueChanged(value: Int)
    }
}
